### PR TITLE
 Fix [has_attack]type= cause infinite recursion when called from [damage_type] which concerns the filtered attack.[1.18]

### PR DESCRIPTION
**Pull request by newfrenchy83**
When for example we have [damage_type][filter_self][has_attack]type=pierce applied to one or any pierce type attack and the unit does not have an unaffected pierce attack, the change in type triggers an infinite recursion.

this PR fixes the problem in the same way as for [filter_weapon]

Fix https://github.com/wesnoth/wesnoth/issues/8940

### CODE REVIEW

File: src/units/abilities.cpp [hunk 1]
```diff
@@ -1311,7 +1311,16 @@ namespace { // Helpers for attack_type::special_active()
 			return false;
 		}
 
-		unit_filter ufilt{vconfig(*filter_child)};
+		//'inform' u.matches what called from a weapon special if filter_child contain "has_attack" child.
+		config cfg = *filter_child;
+		if ((*filter_child).has_child("has_attack") ) {
+			config& filter_has_attack = cfg.mandatory_child("has_attack");
+			filter_has_attack["weapon_type"] = tag_name;
```
Comment (stevecotton) [inline src/units/abilities.cpp:1318]: Adding `filter_has_attack["weapon_type"] = "damage_type"` seems confusing, given the variable name I didn't expect the string to be "damage_type". I think `filter_has_attack["bug_8940"] = tag_name` would be more understandable, as it's immediately obvious that it's a workaround.

I don't like the mechanism, and think it's the wrong solution for the master branch. However, assuming that a better fix will be so large that it nothing in this area gets backported from master to 1.18, it makes sense to fix 1.18 this way.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: Name changed

**Review by stevecotton — COMMENTED** [review]
I'm still unsure about this - got a feeling that I'm missing something, but not sure what.

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg [hunk 2]
```diff
@@ -348,3 +350,249 @@
...
+    [/event]
+)}
+
+#####
+# API(s) being tested: [filter_self][has_attack]type= in [damage_type]
+##
+# Actions:
+# Give Alice an ability that replace all damage by arcane if Alice has a blade attack
+# Define events that use filter_attack matching Alice's arcane type.
+# Have Alice attack Bob.
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg]: corrected

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg [hunk 3]
```diff
@@ -348,3 +350,249 @@
...
+##
+# Expected end state:
+# An event triggers when Alice attacks during side 1's turn because type=arcane detected without [has_attack] cause infinite recursion.
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg]: replaced

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg [hunk 4]
```diff
@@ -348,3 +350,249 @@
...
+            [/move]
+            [attack]
+                [source]
+                    x,y=13,4
+                [/source]
+                [destination]
+                    x,y=13,3
+                [/destination]
+            [/attack]
+        [/do_command]
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg:484]: done

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg [hunk 5]
```diff
@@ -348,3 +350,249 @@
...
+        [store_unit]
+            [filter]
+                id=bob
+            [/filter]
+            variable=b
+        [/store_unit]
+        [unstore_unit]
+            variable=a
+            find_vacant=yes
+            x,y=$b.x,$b.y
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg]: renamed and recoded test

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg [hunk 6]
```diff
@@ -348,3 +350,249 @@
...
+# Give Bob one [damage_type] with alternative_type=cold and another with alternative_type=pierce
+# change resistance Bob to blade to -100% and fire to 0%.
+# Give Alice one [damage_type] with alternative_type=fire
+# and change resistance to cold to -100%, pierce to -50% and blade to 0%.
+# Move Alice next to Bob, and have Alice attack Bob.
+##
+# Expected end state:
+# Alice attack with blade and Bob use cold.
+#####
+{GENERIC_UNIT_TEST "damage_secondary_type_test" (
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg]: renamed to damage_two_secondary_types_test

[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/special_damage_type.cfg]: My bad, i duplicated code of test without realise this. Fixed now

**Pull request closed without merging**